### PR TITLE
sql: Drop CHECK constraints on table when column is dropped

### PR DESCRIPTION
**Pull request by ZacharyThomas**
In PostgreSQL, when a column with a CHECK constraint is dropped, the CHECK
constraint is also dropped. This commit drops a column's CHECK constraints
when the column is being dropped.

Example current behavior:

CockroachDB:

> create table a (a int, b int, check (a < b));
> alter table a drop column b;
> insert into a values (1);
pq: column name "b" not found

PostgreSQL:

> create table a (a int, b int, check (a < b));
> alter table a drop column b;
> insert into a values (1);
INSERT 0 1

After this commit, CockroachDB will drop the check (a < b) constraint.

Release note (sql change): Drop COLUMN now drops CHECK constraints for column as well

touches https://github.com/cockroachdb/cockroach/issues/20630

### CODE REVIEW

File: pkg/sql/alter_table.go [hunk 1]
```diff
@@ -306,6 +307,66 @@ func (n *alterTableNode) startExec(params runParams) error {
...
+					if err != nil {
+						return err, false, nil
+					}
+
+					c, ok := v.(*tree.ColumnItem)
+					if !ok {
+						return nil, true, expr
+					}
+
+					foundCol, err := n.tableDesc.FindActiveColumnByName(string(c.ColumnName))
```
Comment (benesch) [inline pkg/sql/alter_table.go]: You should be able to skip this and compare `c.ColumnName` against `col.Name` directly.

File: pkg/sql/schema_changer_test.go [hunk 2]
```diff
@@ -1245,6 +1245,49 @@ func dropIndexSchemaChange(
...
+func TestDropColumn(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	params, _ := tests.CreateTestServerParams()
+
+	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(context.TODO())
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (k INT PRIMARY KEY, v INT CHECK (v >= 0), a INT DEFAULT 0 CHECK (a <= v), b INT default 100 CHECK (b > a) );
```
Comment (benesch) [inline pkg/sql/schema_changer_test.go]: Can you wrap to multiple lines?

```
CREATE DATABASE t;
CREATE TABLE t.test (
  k INT PRIMARY KEY,
  v INT CONSTRAINT check_v CHECK (v >= 0),
  a INT DEFAULT 0 CONSTRAINT check_av CHECK (a <= v),
  b INT DEFAULT 100 CONSTRAINT check_ab CHECK (b > a)
);
```

I've also named the constraints; see below for details.

File: pkg/sql/schema_changer_test.go [hunk 3]
```diff
@@ -1245,6 +1245,49 @@ func dropIndexSchemaChange(
...
+	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+	if len(tableDesc.Checks) != 3 {
+		t.Fatalf("Expected 3 checks but got %d ", len(tableDesc.Checks))
+	}
+
+	// Bulk insert.
+	const maxValue = 5000
+	if err := bulkInsertIntoTable(sqlDB, maxValue); err != nil {
+		t.Fatal(err)
+	}
```
Comment (benesch) [inline pkg/sql/schema_changer_test.go]: I wouldn't bother adding any data to the table! It'll only slow the test down.

File: pkg/sql/schema_changer_test.go [hunk 4]
```diff
@@ -1245,6 +1245,49 @@ func dropIndexSchemaChange(
...
+
+	if _, err := sqlDB.Exec("ALTER TABLE t.test DROP v"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Re-read table descriptor.
+	tableDesc = sqlbase.GetTableDescriptor(kvDB, "t", "test")
+	// Only the check on column c should remain
```
Comment (benesch) [inline pkg/sql/schema_changer_test.go]: I think you renamed column c to something else? In any case, I recommend: "Only check_ab should remain."

File: pkg/sql/schema_changer_test.go [hunk 5]
```diff
@@ -1245,6 +1245,49 @@ func dropIndexSchemaChange(
...
+	if len(tableDesc.Checks) != 1 {
```
Comment (benesch) [inline pkg/sql/schema_changer_test.go]: Additionally assert that only the proper constraint remains (i.e., `check_ab`).

**Review by benesch — COMMENTED** [review]
Wow, this is awesome! Thanks so much @ZacharyThomas. A few comments below.

**Comment by ZacharyThomas** [comment]
Thanks for the comments! Made suggested changes

**Comment by benesch** [comment]
This is good and fast work! I don't have any additional comments. Constraint master @dt or SQL master @knz, do either of you want to take a look?

---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-:-L1xMqDbMIdnSzNODYa0:bpjv9fl)*

File: pkg/sql/schema_changer_test.go [hunk 6]
```diff
@@ -159,7 +159,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 
 func validExpirationTime(expirationTime int64) bool {
 	now := timeutil.Now()
-	return expirationTime > now.Add(sql.SchemaChangeLeaseDuration/2).UnixNano() && expirationTime < now.Add(sql.SchemaChangeLeaseDuration*3/2).UnixNano()
+	return expirationTime > now.Add(sql.SchemaChangeLeaseDuration / 2).UnixNano() && expirationTime < now.Add(sql.SchemaChangeLeaseDuration * 3 / 2).UnixNano()
```
Comment (benesch) [inline pkg/sql/schema_changer_test.go]: I hate the lack of spacing too, but `gofmt`'s word is gospel. I'm afraid you'll have to put this back the way it was for the "lint" build to succeed.

**Comment by knz** [comment]
Yep, the change looks good.
I have a request about code structure, see below, plus a minor grammar nit for the introductory comment.

Please also extend the commit message to outline how the patch achieves what it does, something like "bla bla bla parses the CHECK expression and checks whether the name of the column appears lexically in the expression bla bla bla"

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/alter_table.go, line 311 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-L1xOD2R2pUpNxb5k9QX:-L1xOD2R2pUpNxb5k9QY:bueiw1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e488203b00effe8cc5977b28dbc25285a1d931/pkg/sql/alter_table.go#L311)):*
> ```Go
> 			}
> 
> 			// Drop check constraints which reference column
> ```

"which reference the column."

---

*[pkg/sql/alter_table.go, line 318 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-L1xONjzF_OWEA5JFpqg:-L1xONjzF_OWEA5JFpqh:bmb91hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e488203b00effe8cc5977b28dbc25285a1d931/pkg/sql/alter_table.go#L318)):*
> ```Go
> 			shouldDrop := map[int]bool{}
> 
> 			for i, check := range n.tableDesc.Checks {
> ```

Extract the body loop and what it does into a separate function `exprContainsColumnName`.
Document what this function does.
Add the following text to the explanation: "warning: this logic only works for 'simple' expressions. If/when CHECK expressions are extended to also support subqueries, this logic will need to be amended."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-:-L1xOlh49L7jtTEF8zu2:bvrn6qu)*

File: pkg/sql/schema_changer_test.go (hunk 6 above)
Comment (ZacharyThomas) [reply pkg/sql/schema_changer_test.go]: No problem. Accidentally formatted with the wrong linter before commit.

**Comment by ZacharyThomas** [comment]
Thanks for the comments! Made structural changes, though not sure if this function belongs in alter table. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-:-L1xlI2fS_2yq-SLdOiq:b-hreim5)*

**Comment by knz** [comment]
Yep, this works!
Just a few minor stylistic changes and this is good to go.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/alter_table.go, line 311 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-L1xOD2R2pUpNxb5k9QX:-L1xoPy7TYI9bJV4KJe9:b-nj85fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e488203b00effe8cc5977b28dbc25285a1d931/pkg/sql/alter_table.go#L311)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"which reference the column."
</blockquote></details>

final period still missing at end of sentence.

---

*[pkg/sql/alter_table.go, line 607 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-L1xoZ6rvLU-zvCK_gAX:-L1xoZ6rvLU-zvCK_gAY:bfrry6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c32da0cdc6b56c2459c102c5cf6cbc65fdcab0d/pkg/sql/alter_table.go#L607)):*
> ```Go
> }
> 
> // WARNING: this logic only works for 'simple' expressions.
> ```

The go linter wants documents before the function to have the following structure on the first line:

```
// <functionname> ....
```

So may I suggest:

```
// exprContainsColumnNames determines whether the given column occurs in the given expression.
// It achieves this using a lexical comparison without considering scoping.
// Warning: ....
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-:-L1xowkEEcCZGeemZVut:b-b67780)*

**Comment by ZacharyThomas** [comment]
*[pkg/sql/alter_table.go, line 311 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-L1xOD2R2pUpNxb5k9QX:-L1y3DT2a6bXDUpwOTcu:bmxdljt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e488203b00effe8cc5977b28dbc25285a1d931/pkg/sql/alter_table.go#L311)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

final period still missing at end of sentence.
</blockquote></details>

d'oh. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203)*

**Comment by knz** [comment]
Brilliant!

Thank youuuuuu!!! :lollipop: 

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-:-L1y9kjzkGfPYuHsiWMk:buwomnl)*

**Comment by ZacharyThomas** [comment]
Thanks for the review!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21203#-:-L1yDvawfpTpCTmM_IDB:b-9qdoz3)*

**Pull request merged**